apps/labs/posts/packaging_painpoints_in_2021.md

```diff
@@ -0,0 +1,228 @@
+
+```bash
+$ nix-shell -I nixpkgs=https://github.com/nixos/nixpkgs/archive/986cf21c45b798c97f2424e17b819af3ecf8083e.tar.gz
+ -p python3Packages.numpy
+ python3Packages.tensorflow python3Packages.pytorch python3Packages.cupy python3Packages.dask python3Packages.mxnet
+```
+
+_My solution:_
+
+```bash
+# To create an environment with the most recent version of all mainstream
+# Python array/tensor libraries installed:
+$ conda create -n many-libs python=3.7
+$ conda activate many-libs
+$ conda install cudatoolkit=10.2
+$ pip install numpy torch jax jaxlib tensorflow mxnet cupy-cuda102 dask toolz sparse  # you may need to turn off the resolver if you're on pip >= 20.3.1
+
+# Conda doesn't manage to find a winning combination here; pip has a hard time
+# too and probably not all constraints are satisfied, but nothing crashes
+# and basic tests work as they are supposed to.
+```
+
+## Conclusion
+
+A lot of good things are happening in various places in the Python packaging world. We're in a better situation than several years ago. As the brainstorm content above shows there are challenges as well, but that has always been the case. If I could have just one wish, I'd wish for good program management across and coordination between the PyPA and Conda communities.
+
+This is my blog post though, so I get a few more wishes:
+
+- Merging Conda and Mamba so they share a single installer and UX with two modes ('fast' and 'stable/legacy').
+- Strictly enforced metadata correctness and manylinux compliance for new PyPI uploads. The stricter the better, see [packaging-problems #264](https://github.com/pypa/packaging-problems/issues/264).
+- GPU hardware for conda-forge so it can test its GPU packages in CI.
+- Consolidate the zoo of virtual environment tools (`venv`, `virtualenv`, `poetry`, `pipenv`, `pyenv`, `pyflow` - probably there are even more) so we can actually recommend one to users of our projects.
+- Not having to think about problems around Fortran on Windows ever again.
+
+While I'm at it, let me also link to some of the most informative blog posts on the topic I know of and enjoyed reading over the years:
+
+- [Tarek Ziadé - The fate of Distutils - Pycon Summit + Packaging Sprint detailed report (2010)](https://tarekziade.wordpress.com/2010/03/03/the-fate-of-distutils-pycon-summit-packaging-sprint-detailed-report/)
+- [Travis Oliphant - Why I promote conda (2013)](http://technicaldiscovery.blogspot.com/2013/12/why-i-promote-conda.html)
+- [Jake VanderPlas - Conda: Myths and Misconceptions (2016)](https://jakevdp.github.io/blog/2016/08/25/conda-myths-and-misconceptions/)
+- [Wes McKinney - conda-forge and PyData's CentOS moment (2016)](https://wesmckinney.com/blog/conda-forge-centos-moment/)
+- [Donald Stufft - Powering the Python Package Index (2016)](https://caremad.io/posts/2016/05/powering-pypi/)
+- [Pauli Virtanen - Building Python wheels with Fortran for Windows (2017)](https://pav.iki.fi/blog/January 24, 2021
```

Comment:
@rgommers this isn't online anymore. Would you prefer to replace it with this Wayback Machine link, or remove it?
```suggestion
- [Pauli Virtanen - Building Python wheels with Fortran for Windows (2017)](https://web.archive.org/web/20210616203153/https://pav.iki.fi/blog/2017-10-08/pywingfortran.html)
```

Reply:
replacing with the archive.org link please